pkg/util/gate/custom_resource_gate.go

```diff
@@ -0,0 +1,78 @@
+package gate
+
+import (
+	"os"
+	"strings"
+
+	"github.com/openkruise/kruise/pkg/apis"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+)
+
+const (
+	envCustomResourceEnable = "CUSTOM_RESOURCE_ENABLE"
+)
+
+var (
+	internalScheme  = runtime.NewScheme()
+	discoveryClient discovery.DiscoveryInterface
+)
+
+func init() {
+	_ = apis.AddToScheme(internalScheme)
+	cfg, err := config.GetConfig()
+	if err != nil {
+		klog.Errorf("unable to set up client config: %v", err)
+		os.Exit(1)
+	}
+	discoveryClient = discovery.NewDiscoveryClientForConfigOrDie(cfg)
+}
+
+// ResourceEnabled help runnable check if the custom resource is valid and enabled
+// 1. If this CRD is not found from kueb-apiserver, it is invalid.
+// 2. If 'CUSTOM_RESOURCE_ENABLE' env is not empty and this CRD kind is not in
+func ResourceEnabled(obj runtime.Object) bool {
+	gvk, err := apiutil.GVKForObject(obj, internalScheme)
+	if err != nil {
+		klog.Warningf("custom resource gate not recognized object %T in scheme: %v", obj, err)
+		return false
+	}
+
+	return discoveryEnabled(gvk) && envEnabled(gvk)
+}
+
+func discoveryEnabled(gvk schema.GroupVersionKind) bool {
+	resourceList, err := discoveryClient.ServerResourcesForGroupVersion(gvk.GroupVersion().String())
+	if err != nil {
+		klog.Warningf("custom resource gate not found groupVersionKind %v in discovery: %v", gvk, err)
+		return false
+	}
+
+	for _, r := range resourceList.APIResources {
+		if r.Kind == gvk.Kind {
+			return true
+		}
```

Comment:
Do we need to check group and version as well?

Reply:
There is no need to check group and version, this `resourceList` is got from discovery by `GroupVersion`.

```
discoveryClient.ServerResourcesForGroupVersion(gvk.GroupVersion().String())
```